book_service/views.py

```diff
@@ -1,9 +1,78 @@
-from rest_framework import viewsets
+from drf_spectacular.utils import extend_schema, OpenApiParameter
+from rest_framework import viewsets, status
+from rest_framework.decorators import action
+from rest_framework.permissions import IsAdminUser
+from rest_framework.response import Response
 
 from book_service.models import Book
-from book_service.serializers import BookSerializer
+from book_service.permissions import IsAdminOrReadOnly
+from book_service.serializers import (
+    BookSerializer,
+    BookImageSerializer,
+    BookListSerializer,
+    BookDetailSerializer,
+)
 
 
 class BookViewSet(viewsets.ModelViewSet):
+    """Endpoints of the books in library with basic CRUD operations"""
+
     queryset = Book.objects.all()
     serializer_class = BookSerializer
+    permission_classes = [IsAdminOrReadOnly]
+    search_fields = ["title", "author"]
+    filterset_fields = ["cover"]
+
+    def get_serializer_class(self):
+        if self.action == "list":
+            return BookListSerializer
+
+        if self.action == "retrieve":
+            return BookDetailSerializer
+
+        if self.action == "upload_image":
+            return BookImageSerializer
+
+        return BookSerializer
+
+    @action(
+        methods=["POST"],
+        detail=True,
+        url_path="upload-image",
+        permission_classes=[IsAdminUser],
+    )
+    def upload_image(self, request, pk=None):
+        """Endpoint for uploading image to a specific book"""
+        book = self.get_object()
+        serializer = BookImageSerializer(book, data=request.data, partial=True)
+
+        if serializer.is_valid():
```

Comment:
Here you can write serializer.is_valid(raise_exception=True)
If the serializer is not valid, an exception will be raised automatically, so there is no need to write the final return.

serializer.is_valid(raise_exception=True)
serializer.save()
return Response(serializer.data, status=status.HTTP_200_OK)